src/proxy/mod.rs

```diff
@@ -0,0 +1,158 @@
+    req: Request<Body>,
+    proxy_type: &ProxyType,
+) -> GenericResult<(Request<Body>, PayloadData)> {
+    match proxy_type {
+        ProxyType::Quicknode => {
+            let (req, payload) = parse_body_payload::<QuicknodePayload>(req).await?;
+            Ok((req, PayloadData::Quicknode(payload)))
+        }
+        ProxyType::Moralis => {
+            let (req, payload) = parse_header_payload::<MoralisPayload>(req).await?;
+            Ok((req, PayloadData::Moralis(payload)))
+        }
+    }
+}
+
+pub(crate) async fn proxy(
+    cfg: &AppConfig,
+    req: Request<Body>,
+    remote_addr: &SocketAddr,
+    payload: PayloadData,
+    x_forwarded_for: HeaderValue,
+    proxy_route: &ProxyRoute,
+) -> GenericResult<Response<Body>> {
+    match payload {
+        PayloadData::Quicknode(payload) => {
+            proxy_quicknode(cfg, req, remote_addr, payload, x_forwarded_for, proxy_route).await
+        }
+        PayloadData::Moralis(payload) => {
+            proxy_moralis(cfg, req, remote_addr, payload, x_forwarded_for, proxy_route).await
+        }
+    }
+}
+
+pub(crate) async fn validation_middleware(
+    cfg: &AppConfig,
+    payload: &PayloadData,
+    proxy_route: &ProxyRoute,
+    req_uri: &Uri,
+    remote_addr: &SocketAddr,
+) -> Result<(), StatusCode> {
+    match payload {
+        PayloadData::Quicknode(payload) => {
+            validation_middleware_quicknode(cfg, payload, proxy_route, req_uri, remote_addr).await
+        }
+        PayloadData::Moralis(payload) => {
+            validation_middleware_moralis(cfg, payload, proxy_route, req_uri, remote_addr).await
+        }
+    }
+}
+
+/// Asynchronously parses an HTTP request's body into a specified type `T`. If the request method is `GET`,
+/// the function modifies the request to have an empty body. For other methods, it retains the original body.
+/// The function ensures that the body is not empty before attempting deserialization into the non-optional type `T`.
+async fn parse_body_payload<T>(req: Request<Body>) -> GenericResult<(Request<Body>, T)>
+where
+    T: DeserializeOwned,
+{
+    let (parts, body) = req.into_parts();
+    let body_bytes = hyper::body::to_bytes(body).await?;
+    if body_bytes.is_empty() {
+        return Err("Empty body cannot be deserialized into non-optional type T".into());
+    }
+    let payload: T = serde_json::from_slice(&body_bytes)?;
+    let new_req = Request::from_parts(parts, Body::from(body_bytes));
+    Ok((new_req, payload))
+}
+
+async fn parse_header_payload<T>(req: Request<Body>) -> GenericResult<(Request<Body>, T)>
+where
+    T: DeserializeOwned,
+{
+    let (parts, body) = req.into_parts();
+    let header_value = parts
+        .headers
+        .get(X_AUTH_PAYLOAD)
+        .ok_or("Missing X-Auth-Payload header")?
+        .to_str()?;
+    let payload: T = serde_json::from_str(header_value)?;
+    let new_req = Request::from_parts(parts, body);
+    Ok((new_req, payload))
+}
+
+fn remove_unnecessary_headers(
```

Comment:
I wouldn't call them *unnecessary* headers.

```suggestion
fn remove_hop_by_hop_headers(
```

Reply:
Done https://github.com/KomodoPlatform/komodo-defi-proxy/pull/22/commits/800de2ce437a7a4c029e5a8f0ecf1285dd93dcbb